toolchain/check/check.cpp

```diff
@@ -56,16 +85,214 @@ auto CheckParseTree(SharedValueStores& value_stores,
 
   context.VerifyOnFinish();
 
-  sem_ir.set_has_errors(err_tracker.seen_error());
+  sem_ir.set_has_errors(unit_info.err_tracker.seen_error());
 
 #ifndef NDEBUG
   if (auto verify = sem_ir.Verify(); !verify.ok()) {
     CARBON_FATAL() << sem_ir << "Built invalid semantics IR: " << verify.error()
                    << "\n";
   }
 #endif
+}
+
+// The package and library names, used as map keys
+using ImportKey = std::pair<llvm::StringRef, llvm::StringRef>;
+
+// Returns a key form of the package object.
+static auto GetImportKey(UnitInfo& unit_info, Parse::Tree::Package package)
+    -> ImportKey {
+  auto* stores = unit_info.unit->value_stores;
+  return {package.package_id.is_valid()
+              ? stores->identifiers().Get(package.package_id)
+              : "",
+          package.library_id.is_valid()
+              ? stores->string_literals().Get(package.library_id)
+              : ""};
+}
+
+static constexpr llvm::StringLiteral ExplicitMainName = "Main";
+
+// Marks an import as required on both the source and target file.
+// TODO: When importing without a package name is supported, check that it's
+// used correctly.
+static auto TrackImport(
+    llvm::DenseMap<ImportKey, UnitInfo*>& api_map,
+    llvm::DenseMap<ImportKey, Parse::Node>* explicit_import_map,
+    UnitInfo& unit_info, Parse::Tree::Package import) -> void {
+  auto package_key = GetImportKey(unit_info, import);
+
+  // Specialize the error for imports from `Main`.
+  if (package_key.first == ExplicitMainName) {
+    // Implicit imports will have already warned.
+    if (explicit_import_map) {
+      CARBON_DIAGNOSTIC(ImportMainPackage, Error,
+                        "Cannot import `Main` from other packages.");
+      unit_info.emitter.Emit(import.node, ImportMainPackage);
+    }
+    return;
+  }
+
+  if (explicit_import_map) {
+    // Check for redundant imports.
+    if (auto [insert_it, success] =
+            explicit_import_map->insert({package_key, import.node});
+        !success) {
+      CARBON_DIAGNOSTIC(RepeatedImport, Error,
+                        "Library imported more than once.");
+      CARBON_DIAGNOSTIC(FirstImported, Note, "First import here.");
+      unit_info.emitter.Build(import.node, RepeatedImport)
+          .Note(insert_it->second, FirstImported)
+          .Emit();
+      return;
+    }
+
+    // Check for explicit imports of the same library.
```

Comment:
I think it'd be useful for this comment to say that the ID comparison here is OK because the package and import directives are from the same file.